### ISSUE TITLE
segmentation colors look incorrect, when chosen from terminology 

### ISSUE THREAD

**Issue by muratmaga** [Slicer/Slicer#8069]
This is with today's preview build.

1. Load MRHead and start segmenting.
2. Use threshold to create a segment, and double click to change it is name to Skin from terminology selector
3. Note that there are now two colors. Original segment color (green), and then another color box that shows the color specified in terminology (brown).
4. Create another segment, and choose another term to see that it contiuues to do the same thing. 

The actual displayed segment color is the color on the left (original segment color, not the terminology one)

![image](https://github.com/user-attachments/assets/23131083-3e39-47c5-8e9c-22a7ee594b8d)

@jcfr  @cpinter @lassoan

**Comment by cpinter** [comment]
I tried to reproduce it with r32942 / 861c8b2 and r33123 / 90861e3 (very latest) and the decoration role did not show up in the name column. I tried this on Windows, @muratmaga told me he used Linux, although there should not be a difference.

**Comment by muratmaga** [comment]
This is with r33130 by the way, I think your revisions are older.

**Comment by muratmaga** [comment]
Session start time .......: 20241128_154733
Slicer version ...........: 5.7.0-2024-11-26 (revision 33130 / c5e8d83) linux-amd64 - installed release
Operating system .........: Linux / 6.8.0-49-generic / #49~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Nov  6 17:42:15 UTC 2 / UTF-8 -

**Comment by cpinter** [comment]
OK I install it on Ubuntu and see what happens.

**Comment by muratmaga** [comment]
This was yesterdays (11/27's) build https://download.slicer.org/?offset=-1

**Comment by muratmaga** [comment]
It reproduces with today's linux preview as well.

**Comment by cpinter** [comment]
I double checked the description and this DOES reproduce on windows. My mistake was not double-clicking the name but the color.

My guess is that it is due to this commit, where the `ColorRole` was changed to `Qt::DecorationRole` at some places
https://github.com/Slicer/Slicer/commit/50612e7ca2e5a26858029a93f23005d7e1fc082b#diff-6c0329496f9e1e52921d60a5363b23b64c84e9c6a552996ff838ffa783eaa28cR163
I used the custom `ColorRole` just so that this swatch does not show up in the cell as the decoration role does.

**Comment by cpinter** [comment]
I tried simply replacing the decoration roles with color roles in the terminology delegate, and it fixed the issue of the segment table reported in this issue, but broke again the color picking in subject hierarchy. I can look in depth tomorrow @muratmaga if that is fine.

Or @Punzo or @lassoan if you know the solution off the bat please let us know. I'd probably start with adding a flag to the delegate and create it differently from the segments table and SH.

**Comment by jcfr** [comment]
As soon as we confirm that switching `Qt::DecorationRole` to the custom `ColorRole`, it will be easy to fix the regression and then update the morphocloud provisioning script to use the updated version.

**Comment by jcfr** [comment]
@cpinter  Thanks for your help looking into this 🙏

**Comment by muratmaga** [comment]
> I tried simply replacing the decoration roles with color roles in the terminology delegate, and it fixed the issue of the segment table reported in this issue, but broke again the color picking in subject hierarchy. I can look in depth tomorrow @muratmaga if that is fine.

I think as long as the fix is incorporated by Saturday morning it is going to be fine, but I will defer the @jcfr since he is the one that needs to change the MorphoCloud install scripts...

**Comment by jcfr** [comment]
> I think as long as the fix is incorporated by Saturday morning it is going to be fine

Indeed, I will be able to update the MorphoCloud infrastructure on Saturday 🚀  (likely later in the evening)

**Comment by Punzo** [comment]
> Or @Punzo or @lassoan if you know the solution off the bat please let us know. I'd probably start with adding a flag to the delegate and create it differently from the segments table and SH.

No sorry. A flag to the delegate class looks good. Not sure if there is a simple way to refactor.

**Comment by cpinter** [comment]
I created a PR with the fixes: https://github.com/Slicer/Slicer/pull/8071

One problem besides the one mentioned in this issue was that Subject hierarchy now didn't allow terminology editing at all, because the terminology delegate set to the color column in subject hierarchy had the default `UseTerminologySelectorCallback`, always returning `false`. In order to allow SH to edit terminologies this callback was created, which necessitated the addition of a new option in Application Settings/Subject hierarchy.

The second commit fixes the problem in this issue. I ended up indeed adding a flag in the delegate to determine whether it is set to a color column or name column.

The video in the PR shows how all this works.
